### PR TITLE
Jsonize vehicle parts requiring other vehicle parts to be installed

### PR DESCRIPTION
**Pull request by Hirmuolio**
#### Summary

```SUMMARY: Infrastructure "Jsonize vehicle parts requiring other vehicle parts"```

#### Purpose of change

Some vehicle parts require other vehicle parts before they can be installed. For example to install turret on a vehicle the tile must already have a turret mount.

This is controlled by flags of which the effect is hardcoded.

Install and uninstall requirements are also defined separately in different places leading to small bugs where install and uninstall requirements are different.

#### Describe the solution
The flags in json have a new property `requires_flag`.

If the flag has `requires_flag` property then this flag means that this part requires that flag to be installed.

If another part is blocking the removal of a part due to it requiring that another part you will see a "Remove the attached [part name]" message. 
If multiple parts are blocking the removal only one of them is displayed.

This system will also work for parts that require more than one part. Just add multiple flags on the vehicle part.  

#### Describe alternatives you've considered

Some of the "this part does something" flags could be separated from "this part requires that other part" flag. For example `ALTERNATOR` flag shouldn't require engine, instead the alternator part should have separate `NEEDS_E_ENGINE` flag for the part requirement.

`requires_flag` is generic name but the system applies to only vehicle parts. The same property could probably be expanded to other things (gunmods, cloth modifications, buildings).

It is possible to destroy the vehicle a vehicle part that is required by another vehicle part without destroying the other vehicle part. For example you can destroy the wheel hub withotu destroying the wheel that is connected on it. The connected part should fall off in these cases but that won't be done here.

#### Testing

Seems to work just fine.

### CODE REVIEW

File: src/flag.h [hunk 1]
```diff
@@ -35,6 +35,14 @@ class json_flag
             return craft_inherit_;
         }
 
+        /** Requires this flag to be installed on vehicle */
+        std::string requires_flag() const {
+            return requires_flag_;
+        }
+        std::string remove_message() const {
+            return cant_remove_;
```
Comment (Qrox) [inline src/flag.h]:
```suggestion
            return cant_remove_.translated();
```

File: src/flag.h [hunk 2]
```diff
@@ -51,6 +59,8 @@ class json_flag
         std::set<std::string> conflicts_;
         bool inherit_ = true;
         bool craft_inherit_ = false;
+        std::string requires_flag_;
+        std::string cant_remove_;
```
Comment (Qrox) [inline src/flag.h]:
```suggestion
        translation cant_remove_;
```

File: data/json/vehicleparts/vp_flags.json [hunk 3]
```diff
@@ -3,7 +3,9 @@
     "id": "ALTERNATOR",
     "type": "json_flag",
     "context": [ "vehicle_part" ],
-    "info": "An alternator.  When mounted on a gasoline or diesel engine and the engine is on, will produce electrical power that can be stored in a battery."
+    "info": "An alternator.  When mounted on a gasoline or diesel engine and the engine is on, will produce electrical power that can be stored in a battery.",
+    "requires_flag": "ENGINE",
+    "cant_remove": "Remove attached alternator first."
```
Comment (Qrox) [inline data/json/vehicleparts/vp_flags.json]: You also need to update `lang/extract_json_strings.py` to extract this string for translation.

File: src/veh_type.h [hunk 4]
```diff
@@ -331,6 +331,9 @@ class vpart_info
         // Display order in vehicle interact display
         int list_order;
 
+        std::set<std::string> get_flags() const {
```
Comment (BevapDin) [inline src/veh_type.h]:
```suggestion
        const std::set<std::string> &get_flags() const {
```
No need to copy the set.

File: src/vehicle.cpp [hunk 5]
```diff
@@ -1344,111 +1345,23 @@ bool vehicle::can_mount( const point &dp, const vpart_id &id ) const
...
-    // Security system must be installed on controls
-    if( part.has_flag( "ON_CONTROLS" ) ) {
-        bool anchor_found = false;
-        for( auto it : parts_in_square ) {
-            if( part_info( it ).has_flag( "CONTROLS" ) ) {
-                anchor_found = true;
+    // Check all the flags of the part to see if they require other flags
+    // If other flags are required check if those flags are present
+    for( const std::string flag : part.get_flags() ) {
+        if( json_flag::get( flag ).requires_flag() != "" ) {
```
Comment (BevapDin) [inline src/vehicle.cpp]:
```suggestion
        if( !json_flag::get( flag ).requires_flag().empty() ) {
```
But better would be to change `requires_flag_` to a `cata::optional<std::string>`.

File: src/vehicle.cpp [hunk 6]
```diff
@@ -1573,24 +1406,23 @@ bool vehicle::can_unmount( const int p, std::string &reason ) const
...
-            std::make_tuple( "WHEEL_MOUNT_LIGHT", "NEEDS_WHEEL_MOUNT_LIGHT", translate_marker( "Remove attached wheel first." ) ),
-            std::make_tuple( "WHEEL_MOUNT_MEDIUM", "NEEDS_WHEEL_MOUNT_MEDIUM", translate_marker( "Remove attached wheel first." ) ),
-            std::make_tuple( "WHEEL_MOUNT_HEAVY", "NEEDS_WHEEL_MOUNT_HEAVY", translate_marker( "Remove attached wheel first." ) )
+    // Find all the flags on parts in this tile that require other flags
+    const point pt = parts[p].mount;
+    std::vector<int> parts_here = parts_at_relative( pt, false );
+
+    std::set<std::string> flags_that_require;
+    for( auto &elem : parts_here ) {
+        for( const std::string flag : part_info( elem ).get_flags() ) {
```
Comment (BevapDin) [inline src/vehicle.cpp]:
```suggestion
        for( const std::string &flag : part_info( elem ).get_flags() ) {
```

File: data/json/vehicleparts/vp_flags.json (hunk 3 above)
Comment (Hirmuolio) [reply data/json/vehicleparts/vp_flags.json]: Do I just add

 ```
    if "cant_remove" in item:
       writestr(outfile, item["cant_remove"], **kwargs)
       wrote = True
```

into `def extract(item, infilename):`?

Comment (Qrox) [reply data/json/vehicleparts/vp_flags.json]: I think so. You can run `lang/update_pot.sh` to see if it correctly extracts these strings.

Comment (Hirmuolio) [reply data/json/vehicleparts/vp_flags.json]: It should add them to the po files in `\lang\po`?
Either it is not working or I don't know how it should be used. Maybe both.

Comment (AMurkin) [reply data/json/vehicleparts/vp_flags.json]: `lang/update_pot.sh` updates lang/po/cataclysm-dda.pot.
You can run `lang/merge_po.sh` after that to update `lang/po/*.po`.
See https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/TRANSLATING.md#maintainers

**Pull request merged**